drivers/at_cmd/at_cmd_interface.c

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#define LOG_LEVEL CONFIG_LOG_DEFAULT_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(at_cmd_interface);
+
+#include <zephyr.h>
+#include <stdio.h>
+#include <net/socket.h>
+#include <init.h>
+#include <bsd_limits.h>
+
+#include <at_cmd_interface.h>
+
+#define RESPONSE_MAX_LEN	CONFIG_AT_CMD_RESPONSE_MAX_LEN
+#define THREAD_PRIORITY		K_PRIO_PREEMPT(CONFIG_AT_CMD_THREAD_PRIO)
+
+#define AT_CMD_OK_STR    "OK"
+#define AT_CMD_ERROR_STR "ERROR"
+#define AT_CMD_CMS_STR   "+CMS:"
+#define AT_CMD_CME_STR   "+CME:"
+
+static int common_socket_fd;
+static struct k_thread socket_thread;
+static K_THREAD_STACK_DEFINE(socket_thread_stack, 2048);
+
+K_MSGQ_DEFINE(return_code_msq, sizeof(enum at_cmd_return_code), 2, 4);
+static K_SEM_DEFINE(cmd_pending, 1, 1);
+
+static at_cmd_interface_handler_t received_data_handler;
+
+static int  last_error;
+
+static enum at_cmd_return_code get_return_code(char *buf)
+{
+	enum at_cmd_return_code return_code = AT_CMD_NOT_RETURN_CODE;
+	char *tmpstr = NULL;
+
+	do {
+		tmpstr = strstr(buf, AT_CMD_OK_STR);
+		if (tmpstr) {
+			return_code = AT_CMD_OK;
+			break;
+		}
+
+		tmpstr = strstr(buf, AT_CMD_ERROR_STR);
+		if (tmpstr) {
+			return_code = AT_CMD_ERROR;
+			break;
+		}
+
+		tmpstr = strstr(buf, AT_CMD_CMS_STR);
+		if (tmpstr) {
+			return_code = AT_CMD_CMS;
+			break;
+		}
+
+		tmpstr = strstr(buf, AT_CMD_CME_STR);
+		if (tmpstr) {
+			return_code = AT_CMD_CME;
+			break;
+		}
+	} while (0);
+
+	if ((return_code == AT_CMD_CMS) || (return_code == AT_CMD_CME)) {
+		char tmpstr[5];
+
+		memcpy(tmpstr, &buf[strlen(AT_CMD_CMS_STR)], 3);
+		last_error = atoi(tmpstr);
+	} else {
+		last_error = 0;
+	}
+
+	if (tmpstr) {
+		buf[tmpstr - buf] = '\0';
+	}
+
+	return return_code;
+}
+
+static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
+{
+	int bytes_read;
+	char buf[RESPONSE_MAX_LEN];
+	enum at_cmd_return_code return_code;
+
+	ARG_UNUSED(arg1);
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+
+	LOG_DBG("Socket thread started");
+
+	for (;;) {
+		bytes_read = recv(common_socket_fd, buf, sizeof(buf), MSG_DONTWAIT);
```

Comment:
Can we not wait in `recv`, instead of sleeping below?

Reply:
The sleep is a hack because the recv function will spin-loop if I tell it to wait until it get new data. I will change the implementation as soon as I know @stig-bjorlykke is done with the proper implementation.